### ISSUE TITLE
Supports Rename Glyphs parameter

### ISSUE THREAD

**Issue by RosaWagner** [googlefonts/glyphsLib#735]
Some issues are talking about it #498 #126 and this one aims to list all the useful parameters that could be handled from source:

- [ ] Remove glyphs
- [x] Keep glyphs
- [x] Rename glyphs
- [ ] Remove features
- [x] Replace feature
- [ ] Remove classes

We are trying to unroll an update of [cormorant](https://github.com/CatharsisFonts/Cormorant/tree/master/4.%20Glyphs%20Source%20Files) which is creating different families by swapping stylistic sets, it looks like that in sources: 

<img width="759" alt="Capture d’écran 2021-10-21 à 14 43 38" src="https://user-images.githubusercontent.com/12222436/138281810-b3dc049e-47e0-4a23-be3f-153aa0762d39.png">

I see how I could reproduce the remove glyphs and remove feature with fonttools subsetter, but not the Rename Glyphs thing though (although making additional post-export script doesn't help the designer to use fontmake). We made a little test case for this particular parameter in case you want to implement it 😬

[test-swap.zip](https://github.com/googlefonts/glyphsLib/files/7389249/test-swap.zip) Exported result should be a circle.

cc @emmamarichal

**Comment by anthrotype** [comment]
while it'd be nice to support all the myriad of custom parameters eventually, in the meantime I believe you should be able to achieve what you need by using the fonttools subsetter (to remove features and glyphs), and perhaps
also @twardoch's `pyftfeatfreeze` script from https://github.com/twardoch/fonttools-opentype-feature-freezer to get what the "Rename Glyphs" does (set .ss02 glyphs as the default ones).

**Comment by anthrotype** [comment]
at least some of the custom params you mention should be supported already. IIRC, we do support Keep Glyphs and Replace feature. Doesn't Remove Glyphs achieve the same thing as Keep Glyphs (only reversed)? Same for Remove Feature, you can replace it with an empty one..
I'm trying to understand which one is really needed, and which one kind of isn't

**Comment by MariannaPaszkowska** [comment]
➕1 to add this functionality. I tried `pyftfeatfreeze` script, and it seems to change a bit too much in the font. I have a lot of unrelated diffs.

**Comment by anthrotype** [comment]
which one you mean? Rename Glyphs?

**Comment by anthrotype** [comment]
How is Rename Glyphs supposed to work exactly?
I found this thread which may be relevant: https://forum.glyphsapp.com/t/swapping-two-characters-on-generate-using-a-custom-parameter/3582/4

**Comment by RosaWagner** [comment]
Rename glyphs swap variant glyphs. For exemple, in one version of the font you want a.ss01 in base glyph and former /a to become a.ss01.

**Comment by anthrotype** [comment]
this blog post (esp. section "2. Alternate glyphs") deals with Rename/Remove Glyphs: https://glyphsapp.com/learn/switching-shapes

**Comment by simoncozens** [comment]
Remove Feature is also easily implementable as a custom parameter handler. Popping that here until we work out which ones we're actually planning on supporting:

```python
class RemoveFeatureParamHandler(AbstractParamHandler):
    def to_ufo(self, builder, glyphs, ufo):
        tags = glyphs.get_custom_values("Remove Features")
        # List of lists, for some reason
        if not tags:
            return
        for tag in tags[0]:
            ufo._owner.features.text = replace_feature(
                tag, "", ufo._owner.features.text or ""
            )

    def to_glyphs(self, glyphs, ufo):
        pass


register(RemoveFeatureParamHandler())
```

The class-related custom parameters are:

* *Add Class*: Add an OpenType class to the font. The first word must be the class name (without the at sign), followed by a semicolon, and the new class code.
* *Replace Class*: Replaces OpenType class code with custom code. The first word must be the class name (without the at sign), followed by a semicolon, and the new class code. Works only if the class exists in File > Font Info > Features. This is only necessary for manually set up classes. Automatically generated classes update automatically.
* *Remove Classes*: Prevents the export of the OpenType classes mentioned in the list. Useful for removing manually written classes when glyphs are removed from the font through the subsetting parameters. Note that automatic classes are removed automatically at export if the triggering glyphs are not in the font anymore, e.g., because they were removed or renamed with parameters.

**Comment by anthrotype** [comment]
> [Keep Glyphs] should have been implemented as a custom parameter handler on the GSInstance, as we have done with the other similar custom parameters

yes, I agree that's odd. I think it's because it uses the fonttools subsetter and requires an ttf/otf binary font to work with. Doing it at the source level would require us to be able to also subset feature files, which at least at that time wasn't available (not sure now, but Simon probably can).


>  but we decided not to support them

If it's that easy to add Remove Glyphs, feel free to add it.
Same for Remove Features, as you proposed in that snippet above.

I'm ok if we eventually add them all, it's just that there are so many of them and I don't have the time to implement them all. But if others will contribute them, sure!

**Comment by simoncozens** [comment]
> Doing it at the source level would require us to be able to also subset feature files

The Glyphs app does not handle this so intelligently. If you have a manually created feature file which mentions glyphs which are removed using Remove Glyphs, it just fails to export. We can easily achieve parity with that.

**Comment by anthrotype** [comment]
hm ok, though I wonder if somebody is actually relying on the fontmake subset feature via Keep Glyphs parameter actually subsetting the layout tables as well, if we implement the Glyphs.app way they may lose that. So I'd just keep that one as is for the time being. Remove Glyphs we can implement the easy way as you proposed.

**Comment by simoncozens** [comment]
We handle auto features by constructing the feature code dynamically in our feature writers. If the glyphs aren't in the UFO at that point, the feature writers won't see them, so no problem.

**Comment by anthrotype** [comment]
I think Khaled meant Glyphs.app's own auto-features for glyph name suffixed GSUB stuff like ss01, smcp, etc.

**Comment by simoncozens** [comment]
Well, these have a well-defined format, so we can just scrub the appropriate lines out of them. I'll work on it.

**Comment by khaledhosny** [comment]
There is also stuff like this:
<img width="1297" alt="image" src="https://user-images.githubusercontent.com/93914/202178335-c26133ad-96b3-41cf-a8ce-a1f7f075ae5c.png">

**Comment by simoncozens** [comment]
Oh urgh. Yeah, easier just to do this with a subsetter.

**Comment by kalapi** [comment]
Hello folks, this is probably the best place to discuss this. I'm having an issue with the 'Replace Feature' custom parameter. I remember that in an older version of `fontmake` if we added the `replaceFeature` custom parameter under each master in the `Masters` tab, glyphsLib would interpolate the values when generating a VF. 

I have two master-specific `dist` features in a file using the above method. This is what they look like:

Light Master:
```
# Automatic Code End
script dev2;

lookup DEV_dlft_half_singleAdjustment {
# Ka
	pos ka-deva.half.a ka-deva.half.b' <0 16 -40 0> pa-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 16 -40 0> pha-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 22 -30 0> nga-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 22 -32 0> jha-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 22 -30 0> dda-deva;
} DEV_dlft_half_singleAdjustment;
```
ExtraBold master:
```
dist;
# Automatic Code End
script dev2;

lookup DEV_dlft_half_singleAdjustment {
# Ka
	pos ka-deva.half.a ka-deva.half.b' <0 26 -40 0> pa-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 26 -40 0> pha-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 28 -20 0> nga-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 28 -20 0> jha-deva;
	pos ka-deva.half.a ka-deva.half.b' <0 28 -20 0> dda-deva;
} DEV_dlft_half_singleAdjustment;
```
The only difference in the two snippets is the the valueRecord.

I also have an empty `dist` feature under the `Features` tab to 'trigger' the inclusion of the feature and without which the font generates without including the `dist` data.

I'm also getting this error which I cannot decode:

```
fontmake.errors.FontmakeError: In 'Documents/[...]/TestFont.glyphs' -> 'master_ufo/TestFont.designspace': Generating fonts from Designspace failed: 

Couldn't merge the fonts, because some values were different, but should have
been the same. This happened while performing the following operation:
GPOS.table.FeatureList.FeatureRecord[0].Feature.LookupListIndex[0]

The problem is likely to be in Test Font Bold:
Expected to see [0]==4, instead saw 3
```
I've tried running the file through Simon's [debug-merge.py](https://gist.github.com/simoncozens/35fc0f75993673ae38b1f1179c7e35f6) script and didn't get any errors. This was the output:

```
Converting to UFO
Checking compatibility of blwm/mkmk/abvm/mark features
Checking compatibility of LigatureCarets/GlyphClassDefs features
```

Is there a new way to implement this? Thanks in advance!

I'm on:
- `fontmake` version 3.9.0
- `glyphsLib` version 6.7.1
- `fonttools` version 4.53.1
